### PR TITLE
Add more rules for backward

### PR DESCRIPTION
**Pull request by dfalbel**
For 

- log
- exp
- reduce_max
- max
- min

Also added `value_and_gradient()` that computes both the value and the gradient of a function.

I have a minimal MNIST example:

```
remotes::install_github("dfalbel/mnist")

library(dotty)

data(mnist, package = "mnist")

.[x, y, ..] <- mnist
x <- array(x / 256, dim = c(60000, 784))
y <- outer(y, 0:9, \(x, y) as.numeric(x == y))

linear <- function(x, params) {
  out <- x %*% params$W
  out + nv_broadcast_to(params$b, shape = shape(out))
}

init_linear_params <- function(nin, nout) {
  list(
    W = nv_tensor(array(rnorm(nin * nout), dim = c(nin, nout))),
    b = nv_tensor(array(0, dim = c(1, nout)))
  )
}

relu <- function(x) {
  nv_max(x, nv_scalar(0))
}

model <- function(x, params) {
  for (p in params[-length(params)]) {
    x <- linear(x, p)
    x <- relu(x)
  }
  linear(x, params[[length(params)]])
}

init_model_params <- function(hidden = c(784, 512, 256, 10)) {
  params <- list()
  for (i in seq_along(hidden[-1])) {
    params[[i]] <- init_linear_params(hidden[i], hidden[i + 1])
  }
  params
}

loss <- function(x, y, params) {
  logits <- model(x, params)
  z_max <- nv_reduce_max(logits, dims = 2, drop = FALSE)
  logsumexp <- z_max + nv_log(nv_reduce_sum(nv_exp(logits - nv_broadcast_to(z_max, shape(logits))), dims = 2, drop = FALSE))
  correct_logit <- nv_reduce_sum(y * logits, dims = 2, drop = FALSE)
  nv_reduce_mean(-correct_logit + logsumexp, dims = c(1, 2))
}

params <- init_model_params()

# like purrr modify_tree but for multiple lists
pmodify_tree <- function(.l, .f, ...) {
  .f <- purrr::as_mapper(.f, ...)
  if (!length(.l)) stop(".l must contain at least one tree")

  rec <- function(nodes) {
    are_list <- vapply(nodes, is.list, logical(1))

    if (all(are_list)) {
      template <- nodes[[1]]
      purrr::imap(template, function(.x, .name) {
        idx <- if (is.null(names(template))) as.integer(.name) else .name
        rec(lapply(nodes, `[[`, idx))
      })
    } else if (!any(are_list)) {
      do.call(.f, nodes)
    } else {
      stop("Trees must have the same shape (mixed list/non-list at some node).", call. = FALSE)
    }
  }

  rec(.l)
}

step <- jit(donate = c("x", "y", "params"), function(x, y, params) {
  .[l, .[grads]] <- value_and_gradient(loss, wrt = "params")(x, y, params)
  params <- pmodify_tree(list(params, grads), \(p, g) p - nv_scalar(0.01)*g)
  list(l, params)
})

params <- init_model_params()
for (i in 1:1000) {
  ids <- sample(1:60000, size = 64)
  .[bx, by] <- list(nv_tensor(x[ids,]), nv_tensor(y[ids,]))
  .[l, params] <- step(bx, by, params)
  if (i %% 100 == 0) {
    cat("loss: ", as.numeric(as_array(l)), "\n")
  }
}

preds <- jit(model)(nv_tensor(x), params)
preds <- as_array(preds)

table(
  apply(preds, 1, which.max),
  apply(y, 1, which.max)
)
```

### CODE REVIEW

File: R/rules-backward.R [hunk 1]
```diff
@@ -60,6 +60,57 @@ p_pow[["backward"]] <- function(inputs, outputs, grads, .required) {
...
+    mask_lhs <- nvl_convert(nvl_eq(lhs, y), dtype = dtype(grad))
+    mask_rhs <- nvl_convert(nvl_eq(rhs, y), dtype = dtype(grad))
+    count <- nvl_add(mask_lhs, mask_rhs)
+  }
+
+  list(
+    if (.required[[1L]]) nvl_div(nvl_mul(grad, mask_lhs), count),
+    if (.required[[2L]]) nvl_div(nvl_mul(grad, mask_rhs), count)
+  )
+}
```
Comment (sebffischer) [inline R/rules-backward.R]:
```suggestion
```
